### PR TITLE
[cpp worker] support pass by reference on cluster mode

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?

- support pass by reference on cluster mode

## Related issue number

https://github.com/ray-project/ray/issues/11388

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by AmeerHajAli** [comment]
Thanks @SongGuyang .
Will test it and get back to you as soon as possible!

File: cpp/include/ray/api/actor_creator.h [hunk 1]
```diff
@@ -29,8 +29,8 @@ ActorCreator<ActorType>::ActorCreator() {}
 
 template <typename ActorType>
 ActorCreator<ActorType>::ActorCreator(RayRuntime *runtime, RemoteFunctionPtrHolder ptr,
-                                      std::shared_ptr<msgpack::sbuffer> args)
-    : runtime_(runtime), ptr_(ptr), args_(args) {}
+                                      std::vector<std::unique_ptr<::ray::TaskArg>> args)
```
Comment (kfstorm) [inline cpp/include/ray/api/actor_creator.h:32]: Why not use a reference here?

File: cpp/src/ray/runtime/task/task_executor.cc [hunk 2]
```diff
@@ -95,34 +88,57 @@ Status TaskExecutor::ExecuteTask(
...
+      auto arg = runtime->Get(task_spec.ArgId(i));
+      auto memory_buffer = std::make_shared<::ray::LocalMemoryBuffer>(
+          reinterpret_cast<uint8_t *>(arg->data()), arg->size(), true);
+      args_buffer.emplace_back(
+          std::make_shared<RayObject>(memory_buffer, nullptr, std::vector<ObjectID>()));
+    } else {
+      auto memory_buffer = std::make_shared<::ray::LocalMemoryBuffer>(
+          const_cast<uint8_t *>(task_spec.ArgData(i)), task_spec.ArgDataSize(i), true);
+      args_buffer.emplace_back(
+          std::make_shared<RayObject>(memory_buffer, nullptr, std::vector<ObjectID>()));
```
Comment (kfstorm) [inline cpp/src/ray/runtime/task/task_executor.cc:108]: Duplicated code.

File: cpp/include/ray/api/arguments.h [hunk 3]
```diff
@@ -10,75 +12,102 @@ namespace api {
...
   /// Notice ObjectRefClassPrefix should be modified by ObjectRef class name or namespace.
   static const std::string ObjectRefClassPrefix = "N3ray3api9ObjectRef";
   std::string type_name = typeid(arg1).name();
+  Serializer::Serialize(packer, arg1);
+  auto memory_buffer = std::make_shared<::ray::LocalMemoryBuffer>(
+      reinterpret_cast<uint8_t *>(buffer.data()), buffer.size(), true);
   if (type_name.rfind(ObjectRefClassPrefix, 0) == 0) {
-    /// Pass by reference.
-    Serializer::Serialize(packer, true);
+    throw RayException("ObjectRef can not be wrapped");
```
Comment (kfstorm) [inline cpp/include/ray/api/arguments.h:61]: How about using `RAY_CHECK` here?

File: cpp/include/ray/api/actor_creator.h (hunk 1 above)
Comment (SongGuyang) [reply cpp/include/ray/api/actor_creator.h:32]: Good point.

File: cpp/src/ray/runtime/task/task_executor.cc (hunk 2 above)
Comment (SongGuyang) [reply cpp/src/ray/runtime/task/task_executor.cc:108]: Yes

File: cpp/include/ray/api/arguments.h (hunk 3 above)
Comment (SongGuyang) [reply cpp/include/ray/api/arguments.h:61]: LGTM

**Review by pcmoritz — APPROVED** [review]

**Review by AmeerHajAli — DISMISSED** [review]
Thanks for the PR @SongGuyang. I wrote a few tests for this that seem to hang forever. I added more combinations of tasks and actors interchanging references and values. It seems the failures happen when you get the objects in a different order of their remote call or if you do not call `Ray::Get` on an object before it is used by the next remote call e.g., (the code after the comment `/// actor task with args which pass by reference`). I was not able to add commits to this PR because I do not seem to have access to this repo, but here is the `cluster_mode_test.cc` that I am using, can we actually replace the current `cluster_mode_test.cc` with this? The test hangs forever, can you please take a look? Our goal here for this to be merged is to get the new test to pass. Can we please do that?

```

#include <gtest/gtest.h>
#include <ray/api.h>
#include <ray/api/ray_config.h>
#include <ray/experimental/default_worker.h>

using namespace ::ray::api;

/// general function of user code
int Return1() { return 1; }
int Plus1(int x) { return x + 1; }
int Plus(int x, int y) { return x + y; }

/// a class of user code
class Counter {
 public:
  int count;

  Counter(int init) { count = init; }
  static Counter *FactoryCreate() { return new Counter(0); }
  static Counter *FactoryCreate(int init) { return new Counter(init); }
  static Counter *FactoryCreate(int init1, int init2) { return new Counter(init1+init2); }
  /// non static function
  int Plus1() {
    count += 1;
    return count;
  }
  int Add(int x) {
    count += x;
    return count;
  }
};

TEST(RayClusterModeTest, FullTest) {
  /// initialization to cluster mode
  ray::api::RayConfig::GetInstance()->run_mode = RunMode::CLUSTER;
  /// TODO(Guyang Song): add the dynamic library name
  ray::api::RayConfig::GetInstance()->lib_name = "";
  Ray::Init();

  /// put and get object
  auto obj = Ray::Put(12345);
  auto get_result = *(Ray::Get(obj));
  EXPECT_EQ(12345, get_result);

  /// common task without args
  auto task_obj = Ray::Task(Return1).Remote();
  int task_result = *(Ray::Get(task_obj));
  EXPECT_EQ(1, task_result);

  /// common task with args
  task_obj = Ray::Task(Plus1, 5).Remote();
  task_result = *(Ray::Get(task_obj));
  EXPECT_EQ(6, task_result);

  /// actor task without args
  ActorHandle<Counter> actor1 = Ray::Actor(Counter::FactoryCreate).Remote();
  auto actor_object1 = actor1.Task(&Counter::Plus1).Remote();
  int actor_task_result1 = *(Ray::Get(actor_object1));
  EXPECT_EQ(1, actor_task_result1);

  /// actor task with args
  ActorHandle<Counter> actor2 = Ray::Actor(Counter::FactoryCreate, 1).Remote();
  auto actor_object2 = actor2.Task(&Counter::Add, 5).Remote();

  /// actor task with args which pass by reference
  ActorHandle<Counter> actor3 = Ray::Actor(Counter::FactoryCreate, 6).Remote();
  auto actor_object3 = actor3.Task(&Counter::Add, actor_object2).Remote();
  int actor_task_result3 = *(Ray::Get(actor_object3));
  EXPECT_EQ(12, actor_task_result3);
  
  int actor_task_result2 = *(Ray::Get(actor_object2));
  EXPECT_EQ(6, actor_task_result2);
 
  /// general function remote call（args passed by value）
  auto r0 = Ray::Task(Return1).Remote();
  auto r1 = Ray::Task(Plus1, 30).Remote();
  auto r2 = Ray::Task(Plus, 3, 22).Remote();

  int result1 = *(Ray::Get(r1));
  int result0 = *(Ray::Get(r0));
  int result2 = *(Ray::Get(r2));
  EXPECT_EQ(result0, 1);
  EXPECT_EQ(result1, 31); 
  EXPECT_EQ(result2, 25);

  /// general function remote call（args passed by reference）
  auto r3 = Ray::Task(Return1).Remote();
  auto r4 = Ray::Task(Plus1, r3).Remote();
  auto r5 = Ray::Task(Plus, r4, r3).Remote();
  auto r6 = Ray::Task(Plus, r4, 10).Remote();

  int result5 = *(Ray::Get(r5));
  int result4 = *(Ray::Get(r4));
  int result6 = *(Ray::Get(r6));
  //int result3 = *(Ray::Get(r3));
  EXPECT_EQ(result0, 1);
  EXPECT_EQ(result4, 2);
  EXPECT_EQ(result5, 3);
  EXPECT_EQ(result6, 12); 
 
  /// create actor and actor function remote call with args passed by value
  ActorHandle<Counter> actor4 = Ray::Actor(Counter::FactoryCreate, 10).Remote();
  auto r7 = actor4.Task(&Counter::Add, 5).Remote();
  auto r8 = actor4.Task(&Counter::Add, 1).Remote();
  auto r9 = actor4.Task(&Counter::Add, 3).Remote();
  auto r10 = actor4.Task(&Counter::Add, 8).Remote();

  int result7 = *(Ray::Get(r7));
  int result8 = *(Ray::Get(r8));
  int result9 = *(Ray::Get(r9));
  int result10 = *(Ray::Get(r10));
  EXPECT_EQ(result7, 15);
  EXPECT_EQ(result8, 16);
  EXPECT_EQ(result9, 19);
  EXPECT_EQ(result10, 27); 

  /// create actor and task function remote call with args passed by reference
  ActorHandle<Counter> actor5 = Ray::Actor(Counter::FactoryCreate, r10).Remote();
  auto r11 = actor5.Task(&Counter::Add, r0).Remote();
  auto r12 = actor5.Task(&Counter::Add, r11).Remote();
  auto r13 = actor5.Task(&Counter::Add, r10).Remote();
  auto r14 = actor5.Task(&Counter::Add, r13).Remote();
  auto r15 = Ray::Task(Plus, r0, r11).Remote();
  auto r16 = Ray::Task(Plus1, r15).Remote();

  int result12 = *(Ray::Get(r12));
  int result14 = *(Ray::Get(r14));
  int result11 = *(Ray::Get(r11));
  int result13 = *(Ray::Get(r13));
  int result16 = *(Ray::Get(r16));
  int result15 = *(Ray::Get(r15));
  
  EXPECT_EQ(result11, 28);
  EXPECT_EQ(result12, 56);
  EXPECT_EQ(result13, 83);
  EXPECT_EQ(result14, 166); 
  EXPECT_EQ(result15, 29); 
  EXPECT_EQ(result16, 30); 
 
  Ray::Shutdown();
}

/// TODO(Guyang Song): Separate default worker from this test.
/// Currently, we compile `default_worker` and `cluster_mode_test` in one single binary,
/// to work around a symbol conflicting issue.
/// This is the main function of the binary, and we use the `is_default_worker` arg to
/// tell if this binary is used as `default_worker` or `cluster_mode_test`.
int main(int argc, char **argv) {
  const char *default_worker_magic = "is_default_worker";
  /// `is_default_worker` is the last arg of `argv`
  if (argc > 1 &&
      memcmp(argv[argc - 1], default_worker_magic, strlen(default_worker_magic)) == 0) {
    default_worker_main(argc, argv);
    return 0;
  }
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

```

**Comment by AmeerHajAli** [comment]
@ericl , I opened an issue here to fix the remaining issue in leaking objects: https://github.com/ray-project/ray/issues/12041

We can merge this for now.

**Review by ericl — APPROVED** [review]

**Pull request merged**